File: cashplot/transactions.py
import re
from typing import NamedTuple
import datetime
from decimal import Decimal
from cashplot.util import *
from cashplot.consts import *

class TransactionBase(EqHash, Repr):
    """A transaction without a category assigned."""

    def __init__(self, date, counter_name, counter_account, description, change):
        assert type(date) is datetime.date
        assert type(counter_name) is str
        assert type(counter_account) is str
        assert type(description) is str
        assert type(change) is Decimal

        self.date = date
        self.counter_name = counter_name
        self.counter_account = counter_account
        self.description = description
        self.change = change


class Transaction(TransactionBase):
    """A transaction with a category assigned."""

    def __init__(self, category, date, counter_name, counter_account, description, change):
        super().__init__(date, counter_name, counter_account, description, change)
        assert type(category) is str
        self.category = category


class TransactionBalance(Transaction):
    """A transaction combined with changes and balances for each account."""

    def __init__(self, category, date, counter_name, counter_account, description, change, changes, balances):
        super().__init__(category, date, counter_name, counter_account, description, change)
        assert type(changes) is dict
        for account_name in changes.keys():
            assert type(changes[account_name]) is Decimal
        assert type(balances) is dict
        for account_name in balances.keys():
            assert type(balances[account_name]) is Decimal
        self.changes = changes
        self.balances = balances


def categorize(transactions, config):
    """Categorize transactions based on matching rules in the configuration."""
    def categorize_transaction(transaction):
        for match_rule in config.match_rules:
            category = match_rule[0]
            match = match_rule[1]
            if not ('counter_name' in match and re.search(match['counter_name'], transaction.counter_name) is None
                    or 'counter_account' in match and re.search(match['counter_account'], transaction.counter_account) is None
                    or 'description' in match and re.search(match['description'], transaction.description) is None):
                return Transaction(category, transaction.date, transaction.counter_name, transaction.counter_account, transaction.description, transaction.change)
        raise LookupError(
            "No rule matching transaction with date " + str(transaction.date))
    return list(map(categorize_transaction, transactions))


def transaction_balances(transactions, config):
    """Calculate the change and balance for every account on each transaction."""
    def zero_accounts():
        accounts = {MAIN_ACCOUNT_NAME: Decimal(0)}
        for account_name in config.savings_accounts:
            accounts[account_name] = Decimal(0)
        return accounts

    def calculate_net_change(changes):
        net_change = Decimal(0)
        for account_name in changes.keys():
            if account_name not in config.net_ignore_accounts:
                net_change += changes[account_name]
        return net_change

    old_balances = config.starting_balances
    tr_balances = []

    for transaction in transactions:
        changes = zero_accounts()

        changes[MAIN_ACCOUNT_NAME] = transaction.change
        if transaction.category in config.savings_accounts:
            changes[transaction.category] = -transaction.change

        balances = zero_accounts()
        for account_name in old_balances.keys():
            balances[account_name] = old_balances[account_name] + \
                changes[account_name]

        transaction_changes = {
            **changes, **{NET_ACCOUNT_NAME: calculate_net_change(changes)}}
        transaction_balances = {
            **balances, **{NET_ACCOUNT_NAME: calculate_net_change(balances)}}
        tr_balances.append(TransactionBalance(transaction.category, transaction.date, transaction.counter_name,
                                              transaction.counter_account, transaction.description, transaction.change, transaction_changes, transaction_balances))

        old_balances = balances
    return tr_balances

File: src/js/lib/transactions.js
/** @module */
const assert = require("nanoassert");
const utils = require("./utils");

const NET_ACCOUNT_NAME = "Net";

/**
 * A single transaction. May have a category assigned.
 */
class Transaction {
  /**
   * Creates a new Transaction instance.
   * @param {Date} date - The date on which the transaction was made.
   * @param {string} counterName - The name of the counter-account.
   * @param {string} counterAccount - The counter-account.
   * @param {string} description - Any additional information about the
   *                               transaction itself.
   * @param {number} change - The amount that was transferred. Negative if sent
   *                          from the main account, positive if received on
   *                          the main account.
   * @param {string} [category] - Category name.
   */
  constructor(
    date,
    counterName,
    counterAccount,
    description,
    change,
    category = null
  ) {
    assert(utils.isValidDate(date));
    assert(typeof counterName === "string");
    assert(typeof counterAccount === "string");
    assert(typeof description === "string");
    assert(Number.isFinite(change));
    if (category !== null) {
      assert(typeof category === "string");
    }

    this.date = date;
    this.counterName = counterName;
    this.counterAccount = counterAccount;
    this.description = description;
    this.change = change;
    this.category = category;
  }

  /**
   * Categorizes this transaction based on matching rules in the configuration.
   * Throws an Error if no categories match.
   *
   * @param {Array<Category>} categories - Category matching rules (see Parameters).
   * @return {Transaction} The categorized transaction.
   */
  categorize(categories) {
    assert(Array.isArray(categories));
    let categoryName = null;
    for (const category of categories) {
      if (
        utils.match(this.description, category.descriptionPattern) &&
        utils.match(
          this.counterName + " " + this.counterAccount,
          category.counterAccountPattern
        )
      ) {
        categoryName = category.name;
        break;
      }
    }

    if (categoryName === null) {
      throw new Error("No categories matched.");
    } else {
      return new Transaction(
        this.date,
        this.counterName,
        this.counterAccount,
        this.description,
        this.change,
        categoryName
      );
    }
  }
}

/**
 * A transaction combined with changes and balances for each account.
 */
class TransactionBalance {
  /**
   * Creates a new TransactionBalance instance.
   * @param {Transaction} transaction - The transaction.
   * @param {Object<number>} changes - A collection of account name => balance
   *                                   change resulting from the transaction.
   * @param {Object<number>} balances - A collection of account name => balance
   *                                   resulting from the transaction.
   */
  constructor(transaction, changes, balances) {
    assert(transaction instanceof Transaction);
    assert(typeof changes === "object");
    for (const account of Object.keys(changes)) {
      assert(Number.isFinite(changes[account]));
    }
    assert(typeof balances === "object");
    for (const account of Object.keys(balances)) {
      assert(Number.isFinite(balances[account]));
    }

    this.transaction = transaction;
    this.changes = changes;
    this.balances = balances;
  }
}

/**
 * Calculate the changes and balances for every account on each transaction.
 * In addition, accumulates the changes from the accounts marked with addToNet
 * to the "Net" account.
 *
 * @param {Array<Transaction>} transactions - The categorized transactions.
 *                                            NOTE: Has to be sorted by date
 *                                                  ascending.
 * @param {Array<Account>} accounts - The account names and parameters.
 * @return {Array<TransactionBalance>} - The calculated
 *                                       transactions+changes+balances.
 */
exports.transactionBalances = function (transactions, accounts) {
  assert(Array.isArray(transactions));
  assert(Array.isArray(accounts));
  assert(accounts.length > 0);

  const mainAccountName = accounts[0].name;

  /**
   * @return {Object} An object with the keys set to the account names and values set to 0.
   */
  function zeroAccounts() {
    const balances = {};
    for (const account of accounts) {
      balances[account.name] = 0;
    }
    return balances;
  }

  /**
   * @param {string} accountName - The account name to check.
   * @return {boolean} Whether an account doesn't contribute to the net total.
   */
  function isNetIgnoreAccount(accountName) {
    for (const account of accounts) {
      if (account.name === accountName && !account.addToNet) {
        return true;
      }
    }
    return false;
  }

  /**
   * @param {string} accountName - The account name to check.
   * @return {boolean} Whether an account is a savings account (i.e. not the first account).
   */
  function isSavingsAccount(accountName) {
    for (const account of accounts) {
      if (account.name === accountName) {
        return account.name !== mainAccountName;
      }
    }
    return false;
  }

  /**
   * @param {object} changes - The changes per account (account name => change objects).
   * @return {number} The total of the changes that contribute to the net value.
   *
   */
  function calculateNetChange(changes) {
    let netChange = 0;
    for (const accountName of Object.keys(changes)) {
      if (!isNetIgnoreAccount(accountName)) {
        netChange += changes[accountName];
      }
    }
    return netChange;
  }

  let oldBalances = zeroAccounts();
  for (const account of accounts) {
    oldBalances[account.name] = account.startingBalance;
  }
  const trBalances = [];

  for (const transaction of transactions) {
    const changes = zeroAccounts();

    changes[mainAccountName] = transaction.change;
    if (isSavingsAccount(transaction.category)) {
      changes[transaction.category] = -transaction.change;
    }

    const balances = zeroAccounts();
    for (const accountName of Object.keys(oldBalances)) {
      balances[accountName] = oldBalances[accountName] + changes[accountName];
    }

    // Deep-copy new balances to oldBalances.
    oldBalances = JSON.parse(JSON.stringify(balances));

    changes[NET_ACCOUNT_NAME] = calculateNetChange(changes);
    balances[NET_ACCOUNT_NAME] = calculateNetChange(balances);

    trBalances.push(new TransactionBalance(transaction, changes, balances));
  }

  return trBalances;
};

exports.Transaction = Transaction;
exports.TransactionBalance = TransactionBalance;
exports.NET_ACCOUNT_NAME = NET_ACCOUNT_NAME;
